### PR TITLE
Experimental API #1 - tiny-atom/fusion

### PR DESCRIPTION
**Pull request by KidkArolis**
- action packs
- where each pack is of shape { state, actions }
- all gets/splits inside actions are scoped to that pack
- actions are bound to split and can be called as functions
- exposes a new method atom.fuse(actions) for extending the action packs

This is built purely on top of the tiny-atom/core with no modifications.
It demos a very decent default approach to working with tiny-atom. Whilst
keeping tiny-atom/core available for any alternative strategies.

Most of all, this preserves the combined effect/reducer approach of
tiny-atom, it preserves the custom merge strategy and is compatible
with all of the existing batteries log/devtool/react/preact.

Finally, it's backwards compatible, you should be able to utilise this in
an existing project gradually if no custom evolve was used.

### CODE REVIEW

**Comment by KidkArolis** [comment]
See `examples/future-example` for a full TODO app demo.

But here's the gist of the idea:

```js
// app.js
const createAtom = require('tiny-atom/future')
const merge = require('tiny-atom/deep-merge')
const actions = require('./actions')

const atom = createAtom(actions, { merge })

// possible to dynamically add the actions
atom.fuse(moduleX)
atom.fuse(moduleY)

atom.observe(render)
```

```js
// actions.js
module.exports.todo = {
  state: {
    items: [
      'learn tiny-atom',
      'use tiny-atom',
      'star tiny-atom'
    ],
    input: ''
  },
  actions: {
    add: (get, split, item) => {
      const items = get().items.concat([item])
      split({ items, input: '' })
      split.analytics.track({ type: 'added' })
    },
    remove: (get, split, index) => {
      const items = get().items.filter((item, i) => i !== index)
      split({ items })
      split.analytics.track({ type: 'removed' })
    }
  }
}

module.exports.analytics = {
  state: {
    events: []
  },
  actions: {
    track: (get, split, event) => {
      const events = get().events.concat([event])
      split({ events })
    }
  }
}
```

get's are scoped by default:

```
get() // scoped
get('analytics') // get another slice
get({ root: true }) // get the whole tree
```

splits are scoped as well:

```
split({ text: 'foo' }) // updates this slice
split.onboarding.close() // triggers an action in another pack
split('onboarding.close') // same as above
```

**Comment by KidkArolis** [comment]
I'm contemplating whether we should make this the main api:

```js
const createAtom = require('tiny-atom') // new
const createAtom = require('tiny-atom/core') // existing
```

or alternatively as a new optional api

```js
const createAtom = require('tiny-atom') // existing
const createAtom = require('tiny-atom/packs') //new
```

**Comment by KidkArolis** [comment]
You can try this with `yarn add tiny-atom@beta`

**Comment by KwanMan** [comment]
Im really liking the `split()` and `split.onboarding()` way of calling actions on another part of the state. Why aren’t we doing that for the `get` as well to keep it consistent, `get()`/`get.analytics()` rather than `get()`/`get(‘analytics’)`

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by KwanMan** [comment]
Also not a fan of potentially effectively having 2 “cores”, anyone who stumbles onto this and wants to quickly use it has to effectively choose a core before even writing any code

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by KidkArolis** [comment]
@KwanMan great suggestion! It's now:

```
get() // scoped to the action pack
get.todo()
get.hints()
```

Which is symmetric with:

```
split({ }) // scoped update
split.todo.add()
split.hints.hide()
```

**Comment by KidkArolis** [comment]
I've renamed this from `future` to `orb`.

Tiny atom will remain what it is, which is a tiny flexible open state management solution at:

```js
require('tiny-atom')
```

And if you'd benefit from the namespaced locally scoped get/split composed from multiple action packs, good for your typical UI app - you can also just use the provided tiny-atom wrapper - `orb`:

```js
require('tiny-atom/orb')
```

In other words, your plain tiny-atom would be:

```js
const createAtom = require('tiny-atom')

const atom = createAtom(initialState, evolve, options)
```

But your typical usage might end up being:

```js
const createAtom = require('tiny-atom/orb')
const merge = require('tiny-atom/deep-merge')
const debug = require('tiny-atom/log')

const packs = {
  entities: { state, actions },
  feeds: { state, actions },
  picks: { state, actions }
}
const atom = createAtom(packs, { merge, debug })
```

This latter form - the orb - is basically a good recommended way to use tiny-atom, it's an opinionated approach - but it's _no longer flexible_. And therefore the core library `require('tiny-atom')`, which _is flexible_, is still the main package entry. It's all shipped in the same package, because `tiny-atom` is batteries included - you install it, and it will cater all your state management needs (structured orb, open core, logger, devtools, preact, react, merge helper, maybe more in the future).

And as a bonus - `orb` is created in such a way that you can switch to it from the basic version of tiny-atom in a backwards compatible way.

File: src/orb.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+  atom.fuse = fuse(atom, actions, cache, options)
+  atom.fuse(initialActions)
+  return atom
+}
+
+function createInitialState (state, actions, options) {
+  if (options.merge) {
+    return Object.keys(actions).reduce((state, namespace) => {
+      const pack = actions[namespace]
+      return typeof pack === 'function' ? state : options.merge(state, { [namespace]: pack.state })
```
Comment (KwanMan) [inline src/orb.js]: why would this be a function?

Comment (KidkArolis) [reply src/orb.js]: This is to support the existing projects that already have large action objects:

```
const actions = {
  existingAction1: () =>{},
  existingAction2: () =>{},
  existingAction3: () =>{},

  onboarding: {
    state: {},
    action1: () => {},
    action2: () => {}
  }
}
```

**Comment by KidkArolis** [comment]
Renamed to `tiny-atom/fusion`. Since it's about fusing multiple action packs together. Reads better than `orb`.

File: src/orb.js (hunk 1 above)
Comment (KwanMan) [reply src/orb.js]: so you mean i can call `split.existingAction1()`? if so, which part of the state will that action be working on?

**Comment by KidkArolis** [comment]
Global - same as it is today.

This whole legacy mode can be turned off with options.strict flag. I think
ill turn strict mode on by default.

**Comment by KidkArolis** [comment]
Will merge for history via #33. Closing this one.

**Pull request closed without merging**